File: backend/src/main/java/com/nutritrack/api/model/Meal.java
package com.nutritrack.api.model;

import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "meals")
@Data
@NoArgsConstructor
public class Meal {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false)
    private LocalDate date;

    @Column(name = "meal_type", nullable = false)
    private String mealType; // BREAKFAST, MORNING_SNACK, LUNCH, EVENING_SNACK, DINNER

    @Column(name = "total_calories")
    private Double totalCalories = 0.0;

    @Column(name = "total_protein")
    private Double totalProtein = 0.0;

    @Column(name = "total_carbs")
    private Double totalCarbs = 0.0;

    @Column(name = "total_fat")
    private Double totalFat = 0.0;

    @OneToMany(mappedBy = "meal", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<MealItem> mealItems = new ArrayList<>();

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}

File: backend/src/models/Meal.js
const mongoose = require('mongoose');

const mealItemSchema = new mongoose.Schema({
  food: { type: mongoose.Schema.Types.ObjectId, ref: 'Food', required: true },
  quantity_g: { type: Number, required: true, default: 100 },
  calories: { type: Number, default: 0 },
  protein: { type: Number, default: 0 },
  carbs: { type: Number, default: 0 },
  fat: { type: Number, default: 0 },
});

const mealSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  date: { type: Date, required: true },
  mealType: { type: String, required: true }, // BREAKFAST, MORNING_SNACK, LUNCH, EVENING_SNACK, DINNER
  items: [mealItemSchema],
  totalCalories: { type: Number, default: 0 },
  totalProtein: { type: Number, default: 0 },
  totalCarbs: { type: Number, default: 0 },
  totalFat: { type: Number, default: 0 },
}, { timestamps: true });

module.exports = mongoose.model('Meal', mealSchema);
